useSelect: incrementally subscribe to stores when first selected from

jsnajdr opened the pull request:
Fixes #47145. On every call of `mapSelect`, it looks at the list of stores it selected from and incrementally subscribes to new ones.

A typical lifecycle in a React component looks like this:
1. React mounts the `useSelect` hook, and `useSyncExternalStore` inside it calls the `getValue` function. The `getValue` function will call `mapSelect`, and initialize the `activeStores` variable with the list of selected stores.
2. In an effect, `useSyncExternalStore` calls the `subscribe` function. It subscribes to the `activeStores`.
3. The next call of `useSelect` calls `mapSelect` again, and notes which stores were selected from this time. It calls `updateStores` with this fresh list.
4. `updateStores` will add new stores to the `activeStores` list, so that new `subscribe` calls subscribe to them, too. And will also go through list of `activeSubscriptions` and subscribes to the new stores for each of them.
5. Finally, `useSelect` hook unmounts and all the subscribed stores are unsubscribed from.

The `updateStores` call is done during `getValue`, which is called during render, which is not 100% OK, because it's a side-effect. But we don't have any other opportunity to update the subscriptions.

To make `updateStores` safe, I'm only adding new subscriptions, never removing old ones. If one call of `mapSelect` selects from `store-1` and a later call selects from `store-2`, we keep the subscription to `store-1`, and only add a new subscription to `store-2`. This is more concurrent-mode safe. It could happen that a `useSelect` render would unsubscribe from `store-1` and then the results of that render wouldn't be committed. Because of a higher-priority update, because of suspension, because of error. The subscription to `store-1` would be incorrectly removed.

Cc: @davilera

**How to test:** Covered by unit tests.

jsnajdr committed:
useSelect tests: unify counter store definitions

jsnajdr committed:
useSelect: incrementally subscribe to stores when first selected from

davilera commented:
This goes beyond my understanding of the internals of `useSelect`, but your description looks pretty solid. Thanks for working on this PR, @jsnajdr! Let's see if someone can review and merge it.

tyxla commented:
Thanks for the thorough explanation in the PR, @jsnajdr 👍 

There's a bunch of test failures - any chance they could be related?

jsnajdr commented:
> There's a bunch of test failures - any chance they could be related?

Yes, it's very likely that there's something wrong with the new `useSelect`. I'll need to investigate that before we can think of merging this.

jsnajdr commented:
I'm returning back to this after some time. A mobile unit test is failing where a Button block is created and saved. The `url` attribute is not `undefined`, but an empty string, `""`. There is some subtle bug either in the new `useSelect` implementation, or in the React Native code that works with the blocks store.

jsnajdr commented:
Something goes wrong when there are two selects where the second depends on the result of the first one:
```js
const first = useSelect( select => select( store ).getFirst() );
const second = useSelect( select => first + select( store ).getSecond(), [ first ] );
```
Then, after the `getFirst()` store value changes, it's like the second select callback still uses the old `first` value.

Before this PR, the second select would re-subscribe to the `store`, i.e., unsubscribe and subscribe again, when the dependencies change. That leads to extra calculation of the latest value. But with this PR, `useSelect` detects that the selected stores haven't really changed and doesn't resubscribe. That's smart, but apparently there is some little bug hidden there.

jsnajdr committed:
useSelect: keep lastMapResult and lastMapSelect always in sync

jsnajdr commented:
Well this was quite hard to debug but I think I found the culprit. When there is a `useSelect` with dependencies:
```js
useSelect( ( select ) => {
  return select( blockEditorStore ).getBlockName( clientId );
}, [ clientId ] );
```
we need to correctly handle the transition when the `clientId` changes. There is a new `mapSelect` callback being passed in, with a completely different return value and meaning.

The following was happening here, leading to bad behavior of the editor (crashes, failed tests):
1. There is an internal `useSyncExternalStore( subscribe, getValue )` call, and the `getValue` function is bound to a certain version of the `mapSelect` callback.
2. When there is a store update, React will call this `getValue` callback to get the latest value. The `getValue` callback is saved in the hook's internal state.
3. When `getValue` changes (after a `mapSelect` update, after a dependency update), the new version is saved to internal state in an effect. That means there is a little delay between the render that produced the new `getValue` and the effect that saves it.
4. `getValue` caches the result value, and returns the cached value if current `mapSelect` is same as `lastMapSelect`.
5. `lastMapSelect` is set during render, before the new `getValue` callback is stored into internal state.
6. When a store update comes between render and effect, it still uses the old `getValue`, and will compute and cache a value computed by the old `mapSelect`.
7. But then then following render will use that incorrect cached value because it thinks it was calculated with the new `mapSelect`.

So, it can happen that a render with a new `clientId` will get a value computed using the old `clientId`. That's a complete disaster for the entire editor.

My solution is to set the `lastMapSelect` always in sync with the `lastMapResult`. Then we won't return incorrect cached values. At all times, we know that `lastMapResult` was computed by `lastMapSelect`, they're never out of sync, and we can decide whether we like that value or not.

Before this PR, we were always re-subscribing to all stores on a dependency update. New subscription invalidates the cached value, so the bug was never triggered. Now, after this PR, the code is smarter, it knows that we still read only from `blockEditorStore`, no matter what the `clientId` is, so we don't need to re-subscribe. But it triggered the bug.

At this moment I don't know how to write a test to catch the bug. I suspect that in `act()` testing environment, the right timing might be impossible to reproduce.

jsnajdr commented:
This PR is now green and can be reviewed and merged. FYI @youknowriad @kevin940726

jsnajdr commented:
I have some concerns whether this `useSelect` implementation is really correct. Consider this usage:
```js
const name = useSelect( ( select ) => {
  return select( blockEditorStore ).getBlockName( clientId );
}, [ clientId ] );
```
The component with the hook is mounted, with `clientId="A"`, and is rerendered whenever the `A` block's name changes. Now let's rerender it with `clientId="B"`. There's this sequence of events:
1. The component is rendered, `useSelect` is called with `[ 'B' ]` dependencies, and with a callback whose `clientId` refers to `'B'` in the closure.
2. The `useSelect`'s internal `useSyncExternalStore` returns a new `getValue` callback, bound to the new `mapSelect`, and it will be returning name of block `B`.
3. `useSyncExternalStore` will use this new `getValue` to calculate `B`'s name and return it.
4. The component is rendered with variable `name` having the value of `B` block's name, in sync with the `clientId` prop.
5. An effect is scheduled to store the new `getValue` into the hook's internal state.
6. Now, after render and before the effect, a new store update comes. `useSyncExternalStore`'s listener will be called. The listener will compute the latest value, but it will use the old version of `getValue`, stored in internal state, and bound to `clientId="A"`! If, in this update, `B`'s name has changed and `A`'s name didn't, no component rerender will be triggered. The component will remain rendered with old `B`'s name.

Before this PR, `useSelect` would unsubscribe from the stores and subscribe again when the `clientId` dependency changed. And `useSyncExternalStore` has methods to catch missed updates between the render and the effect that actually establishes the subscription. It will call `getValue` again right after subscription to check if the value has changed. That part is now missing.

I don't have proof or test that this really happens, and I didn't see it happen in production. But the above reasoning tells me that it can happen.

jsnajdr commented:
There is a [`useSyncExternalStoreWithSelector`](https://github.com/facebook/react/blob/main/packages/use-sync-external-store/src/useSyncExternalStoreWithSelector.js) hook, direcly in the React repo, used by `react-redux`. I'll have to study if and how it avoids this "store listener using a stale selector" problem.

jsnajdr committed:
useSelect: improve unit test that tests store updates between render and subscription

jsnajdr committed:
useSelect: add test for store updates after selector change

jsnajdr commented:
> I have some concerns whether this `useSelect` implementation is really correct.

I looked at this more closely today, and tested both our and `react-redux` implementation, and my concerns turn out to be unfounded. If a store update comes in between selector change (render) and effect, then yes, it will use the stale selector, but then in the effect it will always do another `checkIfSnapshotChanged`, this time with the new selector, and the missed update will be processed.

I improved the test cases for this. There was an old test for updates between render and subscribe, and I simplified it, removing class component in favor of a functional one. And added a new test case for the store update between selector change and effect.

Both test cases schedule the insidious store update using `useLayoutEffect`. That ensures it will happen after render and before the normal effect that `useSyncExternalStore` schedules. I learned about this technique in `react-redux` test.

This PR is ready for review and merging once again.

jsnajdr commented:
@youknowriad @kevin940726 Could you please have a look at this one? Seems like the pings went through the cracks.

youknowriad commented on packages/data/src/components/use-select/test/index.js:
```diff
@@ -109,7 +122,7 @@ describe( 'useSelect', () => {
 
 		expect( selectSpyFoo ).toHaveBeenCalledTimes( 2 );
-		expect( selectSpyBar ).toHaveBeenCalledTimes( 2 );
+		expect( selectSpyBar ).toHaveBeenCalledTimes( 1 );
```
So there's a perf improvement here or something?

jsnajdr replied:
Kind of, because when you do a select like this:
```js
const name = useSelect( ( select ) => {
  return select( blockEditorStore ).getBlockName( clientId );
}, [ clientId ] );
```
then before this PR, when the `clientId` dependency changed, `useSelect` unsubscribed from the `blockEditorStore` and then immediately subscribed again. After this PR, this is no longer done, because `useSelect` knows that we're still selecting from this one store, that the set of observed stores didn't change. Only the `clientId` param has changed.

This kind of select is very common in block sidebar and toolbar, as they show info for the currently selected block.

youknowriad reviewed (approved):
To be honest, this is a little bit hard to review but the test changes make sense and are a good indication that this works as intended.

youknowriad commented:
I actually suspect that this PR increased typing performance by approximatively 10%. Is there anything we can do about it.

If we can't do anything about it, we should ask ourselves what's better:

 - conditional useSelects support
 - 10% perf in typing metric

youknowriad commented:
It actually impacted other metrics similarly as well (block select, inserter open, inserter hover)

jsnajdr commented:
> I actually suspect that this PR increased typing performance by approximatively 10%.

The extra work the new version does is tracking the `listeningStores` set on each `mapSelect` call and comparing it to the previous, stored one. Previously these checks were done only when:
- for `useSelect` without dependencies (done mainly by `withSelect`), the checks were never done, the hook never resubscribed.
- for `useSelect` with dependencies, the checks were done only when dependencies changes.

Now, to support conditional selects, we need to check on each store update.

> Is there anything we can do about it.

One thing that could improve the perf again is getting rid of `withSelect` and other `useSelect` calls without specified dependencies. Selecting without dependencies means that the hook can [never use a cached return value](https://github.com/WordPress/gutenberg/blob/ac2d3feb60f8984a1f7cecf27f7d138396982253/packages/data/src/components/use-select/index.js#L124-L129).

> If we can't do anything about it, we should ask ourselves what's better:

That's not an easy tradeoff because we'd be trading performance for correctness. Like the Pentium fdiv bug from 1994 🙂

youknowriad commented:
> That's not an easy tradeoff because we'd be trading performance for correctness. Like the Pentium fdiv bug from 1994

Indeed, because we never supported these conditional selects, personally I'm willing to just document that we don't support them. I don't think it's incorrect, it's just different.
